src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1165,40 +1174,69 @@ def _generate_partition(
-        lower_stride = processed_lower_bound / actual_num_partitions
-        stride = upper_stride - lower_stride
+        upper_stride = (
+            processed_upper_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        lower_stride = (
+            processed_lower_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        preciseStride = upper_stride - lower_stride
+        stride = int(preciseStride)
+
+        lost_num_of_strides = (
+            (preciseStride - decimal.Decimal(stride))
+            * decimal.Decimal(actual_num_partitions)
+            / decimal.Decimal(stride)
+        )
+        lower_bound_with_stride_alignment = processed_lower_bound + int(
+            (lost_num_of_strides / 2 * decimal.Decimal(stride)).quantize(
+                decimal.Decimal("1"), rounding=ROUND_HALF_UP
+            )
+        )
+
+        current_value = lower_bound_with_stride_alignment
 
         partition_queries = []
         for i in range(actual_num_partitions):
-            left = (
-                processed_lower_bound + i * stride
-                if column_type != int
-                else int(processed_lower_bound + i * stride)
+            l_bound = (
+                f"{column} >= '{self._to_external_value(current_value, column_type)}'"
+                if i != 0
+                else ""
             )
-            right = (
-                min(left + stride, processed_upper_bound)
-                if column_type != int
-                else int(min(left + stride, processed_upper_bound))
-            )
-            partition_queries.append(
-                select_query
-                + f" WHERE {column} >= {self._to_external_value(left, column_type)} and {column} {'<=' if right == processed_upper_bound else '<'} {self._to_external_value(right, column_type)}"
+            current_value += stride
+            u_bound = (
+                f"{column} < '{self._to_external_value(current_value, column_type)}'"
+                if i != actual_num_partitions - 1
+                else ""
             )
 
+            if u_bound == "":
+                where_clause = l_bound
+            elif l_bound == "":
+                where_clause = f"{u_bound} OR {column} is null"
+            else:
+                where_clause = f"{l_bound} AND {u_bound}"
+
+            partition_queries.append(select_query + f" WHERE {where_clause}")
+
         return partition_queries
 
-    def _to_internal_value(self, value, column_type):
+    def _to_internal_value(self, value: Union[int, str, float], column_type: DataType):
         if isinstance(column_type, _NumericType):
             return int(value)
+        elif isinstance(column_type, (TimestampType, DateType)):
+            dt = parser.parse(value)
+            return int(dt.replace(tzinfo=pytz.UTC).timestamp())
```

Comment:
log a JIRA for timestamp with tz support for partition column, current impl deviates from pyspark when handling timestamp with tz